File: Amazon/Trees and Graphs/_ Validate_Binary_Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return check_bst(root, nullptr, nullptr);
    }
    
private:
    bool check_bst(TreeNode* root, TreeNode* max, TreeNode* min){
        if(root == nullptr) return true;
        if(max != nullptr and root->val >= max->val) return false;
        else if(min != nullptr and root->val <= min->val) return false;
        
        return check_bst(root->left, root, min) and check_bst(root->right, max, root);
    }
};

File: Amazon/Trees and Graphs/ Validate Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        # def traverse(node, left=float('-inf'), right=float('inf')):
        #     if node is None: return True
        #     if node.val <= left or node.val >= right: return False
        #     if not traverse(node.left, left, node.val): return False
        #     if not traverse(node.right, node.val, right):return False
        #     return True
        left=float('-inf')
        right=float('inf')
        stack = [(root, left, right)]
        while stack:
            node, left, right = stack.pop()
            if not node: continue
            if node.val <= left or node.val >= right: return False
            stack.append((node.left, left, node.val))
            stack.append((node.right, node.val, right))
        return True
        
    